packages/button/index.tsx

```diff
@@ -0,0 +1,135 @@
+import React from 'react'
+import slugify from 'slugify'
+import fragment from './fragment.graphql'
+import classNames from 'classnames'
+import useProductMeta from '@hashicorp/platform-product-meta'
+import InlineSvg from '@hashicorp/react-inline-svg'
+import svgArrowRight from './icons/arrow-right.svg?include'
+import svgExternalLink from './icons/external-link.svg?include'
+import svgCornerRightDown from './icons/corner-right-down.svg?include'
+import svgDownload from './icons/download.svg?include'
+import s from './style.module.css'
+import sTheme from './theme.module.css'
+import useHover from './hooks/use-hover'
+import normalizeButtonTheme from './helpers/normalizeButtonTheme.js'
+import { Size, LinkType, IconObject, Theme, IconProps } from './types'
+
+const linkTypeToIcon = {
+  inbound: svgArrowRight,
+  outbound: svgExternalLink,
+  anchor: svgCornerRightDown,
+  download: svgDownload,
+}
+
+interface ButtonProps {
+  title: string
+  url?: string
+  label?: string
+  external?: boolean
+  theme?: Theme
+  ga_prefix?: string
+  onClick?: React.MouseEventHandler<HTMLAnchorElement> &
+    React.MouseEventHandler<HTMLButtonElement>
+  disabled?: boolean
+  className?: string
+  linkType?: LinkType
+  icon?: IconObject
+  size?: Size
+  [x: string]: any
```

Comment:
@BRKalow whoa, wizardry!

Tried to implement  that for a bit, and honestly having a hard time wrapping my head around how to set this up nicely. Running into various TS errors, some of which I can kind of resolve, others I have no idea.

1. I can't seem to destructure `url` as it doesn't seem to exist on the type, despite using the example above (I'm likely doing something wrong here, just not sure what):

    <img width="490" alt="CleanShot 2021-08-27 at 21 28 15@2x" src="https://user-images.githubusercontent.com/4624598/131202056-b8d2d3e5-6766-442c-99ad-0b5d1fdc7cb8.png">

2. I can't seem to define `title` as a required attribute. This maybe comes down to bad naming on our part - I guess `text` or even just using `children` may have been a better choice, as it wouldn't collide with the HTML `title` attribute? Not sure the benefit vs drawback of having the type annotation for HTML `title` show up when we're actually using that for something else, and when it is in fact required.

    <img width="1060" alt="CleanShot 2021-08-27 at 21 27 55@2x" src="https://user-images.githubusercontent.com/4624598/131202107-76974cfa-5ba4-4392-8211-a9b4be734558.png">

3. Explicitly passing `onClick` seems to make TS very angry. Not sure why this is, I find these tooltip things pretty hard to decipher most of the time. (Should note this seems unrelated to the above issue - this is when I remove any explicit mention of `onClick`, apart from passing it to `Elem`).
 
    <img width="1284" alt="CleanShot 2021-08-27 at 21 28 51@2x" src="https://user-images.githubusercontent.com/4624598/131202134-a3f574de-1f46-4635-90eb-ac64dbc08dc7.png">

4. After doing my best to resolve the above three issues (I still get the `url` error; I remove the explicit `title` definition and that "fixes" it; I stop passing `onClick` explicitly and that "fixes" that error), I still get this pretty wild error (which again, having a hard time deciphering, likely just due to lack of experience w Typescript):

    <img width="1270" alt="CleanShot 2021-08-27 at 21 29 23@2x" src="https://user-images.githubusercontent.com/4624598/131202251-87a2bb0e-d974-4b7f-b210-bb27ab10fe31.png">

Reply:
Also appreciate though that this might just come down to not using Typescript from the start for this component.

I feel like **fully separating** `ButtonButton` and `AnchorButton` (as they are named above) and making it up to the consumer to choose the right component depending on their need for an `onClick` or `href` might be a better move long term (as well as other tweaks such as renaming the `title` prop). Seems like it'd simplify things in terms of the type interface, anyways.

That being said, no matter how we approach trying to fix this, I think it seems like a relatively high LOE with relatively low payoff in the context of this PR:

- If we go manage to figure out a non-breaking interface, as above, that somehow perfectly dynamically toggles between `button` and `anchor` types, I think we'll end up making the ButtonProp type(s) really complex and arguably (IMO) less easy to understand for us and other maintainers (though maybe this is just my newness to TS talking).
- If we go with a bigger refactor, and split the components, this will be a decent amount of work in itself. And perhaps more significantly, it'll add to CSS modules rollout work, which already seems like it'll be pretty significant, due to a smörgåsbord of overrides out there in projects we'll need to deal with.

Given the above hurdles, I think for the sake of preserving the momentum of CSS modules conversion it might make sense for us to go with the `$TSFixMe` as the type for now, per [the note in nextjs-scripts on tracking use of any](https://github.com/hashicorp/nextjs-scripts/blob/613ab00400478cab8a6fe16585a6325bc5ef2890/packages/types/index.d.ts#L21). Going to go that route unless y'all have any objection 🙇

Reply:
Happy to roll with `$TSFixMe` for now to keep things moving, maybe just a TODO with a link to an asana task? This would be a good typing pattern to have nailed down as a team, which we can hopefully implement and then document for future use cases. Maybe something for a hack day 😄

Reply:
👍 Added a comment, and linked to [this new task I created](https://app.asana.com/0/1100423001970639/1200880473915564). Let me know if that looks alright to you 🙇